src/include/sof/coherent.h

```diff
@@ -112,16 +114,29 @@ __must_check static inline struct coherent *coherent_acquire(struct coherent *c,
 
 		c->key = k_spin_lock(&c->lock);
 
+		/*
+		 * FIXME: This is wrong. dcache_invalidate_region() only makes
+		 * sense if we assume, that dirty cache lines might exist for
+		 * this object. But in that case those lines could be written
```

Comment:
No - this is correct. No object should be accessed without acquire first. If cache lines are dirty in this object (via modifying the object without holding the lock) then its game over. 

Btw, for perf, we could be clever and do something like
```diff
diff --git a/rimage b/rimage
index ee5c3e1de..0f64a2018 160000
--- a/rimage
+++ b/rimage
@@ -1 +1 @@
-Subproject commit ee5c3e1deb1436a6e52b653bb523b594511cf330
+Subproject commit 0f64a20183a8db9bc39a4044679a51c1cc334f8f
diff --git a/src/include/sof/coherent.h b/src/include/sof/coherent.h
index bfa6180d9..03ceab7a3 100644
--- a/src/include/sof/coherent.h
+++ b/src/include/sof/coherent.h
@@ -65,7 +65,8 @@ struct coherent {
        uint8_t sleep_allowed;  /* the object will never be acquired or released
                                 * in atomic context */
        uint8_t shared;         /* shared on other non coherent cores */
-       uint16_t core;          /* owner core if not shared */
+       uint8_t core;           /* owner core if not shared */
+       uint8_t last_used_core; /* core that last acquired object */
        struct list_item list;  /* coherent list iteration */
 } __coherent;
 
@@ -112,8 +113,9 @@ __must_check static inline struct coherent *coherent_acquire(struct coherent *c,
 
                c->key = k_spin_lock(&c->lock);
 
-               /* invalidate local copy */
-               dcache_invalidate_region(uncache_to_cache(c), size);
+               /* next user is a different core so WB and INV object */
+               if (c->last_used_core != this_core())
+                       dcache_writeback_invalidate_region(uncache_to_cache(c), size);
        }
 
        /* client can now use cached object safely */
@@ -130,8 +132,8 @@ static inline struct coherent *coherent_release(struct coherent *c, const size_t
        if (c->shared) {
                CHECK_COHERENT_CORE(c);
 
-               /* wtb and inv local data to coherent object */
-               dcache_writeback_invalidate_region(c, size);
+               /* set last user before we release */
+               c->last_used_core = this_core();
 
                /* unlock on uncache alias */
                k_spin_unlock(&cache_to_uncache(c)->lock, cache_to_uncache(c)->key);
```

Reply:
> No - this is correct. No object should be accessed without acquire first. If cache lines are dirty in this object (via modifying the object without holding the lock) then its game over.

@lgirdwood that's the same as what I was trying to say :-) If everybody is following the rules and only accessing the object between the acquire and release, then all is good, and *outside* of that block, when the object isn't acquired, nobody is accessing it via cache, so cache is never dirty, so invalidating it isn't needed. Only if that rule is violated and someone access the object without acquiring it, only then can cache become dirty, but then just we have a bigger problem and just dropping that cache might just as well be wrong.

Reply:
> Btw, for perf, we could be clever and do something like

@lgirdwood I don't think that would work, because with this:
```diff
-               /* wtb and inv local data to coherent object */
-               dcache_writeback_invalidate_region(c, size);
+               /* set last user before we release */
+               c->last_used_core = this_core();
```
we skip writing back cache on *previous* core and this
```diff
-               /* invalidate local copy */
-               dcache_invalidate_region(uncache_to_cache(c), size);
+               /* next user is a different core so WB and INV object */
+               if (c->last_used_core != this_core())
+                       dcache_writeback_invalidate_region(uncache_to_cache(c), size);
```
attempts to write back that cache, but now we're running on the new core... Or am I misunderstanding anything?

Reply:
> @lgirdwood that's the same as what I was trying to say :-) If everybody is following the rules and only accessing the object between the acquire and release, then all is good, and _outside_ of that block, when the object isn't acquired, nobody is accessing it via cache, so cache is never dirty, so invalidating it isn't needed. Only if that rule is violated and someone access the object without acquiring it, only then can cache become dirty, but then just we have a bigger problem and just dropping that cache might just as well be wrong.

Yes, and **we** have to enforce usage of the coherent API for shared objects. The first step is to provide the API though, and then we can address existing users...

Reply:
> @lgirdwood I don't think that would work, because with this:
> 
> ```diff
> -               /* wtb and inv local data to coherent object */
> -               dcache_writeback_invalidate_region(c, size);
> +               /* set last user before we release */
> +               c->last_used_core = this_core();
> ```
> 
> we skip writing back cache on _previous_ core and this
> 
> ```diff
> -               /* invalidate local copy */
> -               dcache_invalidate_region(uncache_to_cache(c), size);
> +               /* next user is a different core so WB and INV object */
> +               if (c->last_used_core != this_core())
> +                       dcache_writeback_invalidate_region(uncache_to_cache(c), size);
> ```
> 
> attempts to write back that cache, but now we're running on the new core... Or am I misunderstanding anything?

Good, find - we just need to WB the struct coherent part only (although this should be done by the atomic unlock).

Reply:
@lgirdwood sorry, still don't understand - why only the coherent header structure? If core A acquired a shared object X, modified it via its cache, didn't write cache back, then core B tries to acquire the same object and sees that it was previously used by a different core - how does it get contents of A's cache for the object?

Reply:
> @lgirdwood sorry, still don't understand - why only the coherent header structure? If core A acquired a shared object X, modified it via its cache, didn't write cache back, then core B tries to acquire the same object and sees that it was previously used by a different core - how does it get contents of A's cache for the object?

Your right, this wont work as the L1$ are all core local. We will need to always WB on release and INV on acquire.

Reply:
The comment can be removed here as ALL users MUST use the API.

Reply:
> Yes, and **we** have to enforce usage of the coherent API for shared objects. The first step is to provide the API though, and then we can address existing users...

This cache invalidation is in `coherent_acquire()` so it is executed when somebody already **is** using the coherent API. And if they're using it correctly, it shouldn't be needed. And if they're using it wrongly - all bets are off.

Reply:
@lyakh can you update the comment, nothing has to be fixed in code - the fix is really that users have to use the API for accessing shared objects.

Reply:
@lgirdwood I think the code below can help only in some specific API-misuse cases. If the API is used correctly, it isn't needed, right? I think we agree about this.
If the API is used incorrectly, I don't think this fixes **all** possible API usage errors. It **could** help for cases when someone only populated cache with the data, i.e. read it bypassing the API via a cached alias. Take this example:
```
struct comp_buffer {
	struct coherent c;

	/* data buffer */
	struct audio_stream stream;
	...
}

struct audio_stream {
	/* runtime data */
	uint32_t size;	/**< Runtime buffer size in bytes (period multiple) */
	...
}

static size_t bad_user(struct comp_buffer *buf)
{
	/* BUG: getting cached alias bypassing the API */
	struct comp_buffer *buf_c = uncache_to_cache(buf);
	/* BUG: reading from object via cached address when not acquired */
	size_t size = buf_c->stream.size; // read access - populates cache but doesn't mark it dirty
	buf_c = buffer_acquire(buf);
```
in that case invalidating cache will indeed prevent data from being overwritten and will save the day. But if instead of the read access line above we have
```
	buf_c->stream.size = 128; // write access: populate cache and mark it dirty
```
then invalidating that line will lose user data.

Now, instead of that cache invalidation we could actually add a trap for API abusers. We could do
```
	coherent_uncached->any_element = ~coherent_uncached->any_element;
	if (memcmp(coherent_uncached, coherent_cached, sizeof(*coherent))) {
		printk("ERROR");
		panic();
	}
	coherent_uncached->any_element = ~coherent_uncached->any_element;
```
and remove that invalidation.

Reply:
> I think the code below can help only in some specific API-misuse cases. If the API is used correctly, it isn't needed, right? I think we agree about this.

I want misuse to fail fast, i.e. misuse to be fixed and not helped. Merging this now as needed for v2.2, we can align this for v2.3